src/plugins/hf3fs/hf3fs_backend.cpp

```diff
@@ -0,0 +1,388 @@
+    }
+
+    out = (nixlBackendMD*) md;
+    return status;
+}
+
+nixl_status_t nixlHf3fsEngine::deregisterMem (nixlBackendMD* meta)
+{
+    nixlHf3fsMetadata *md = (nixlHf3fsMetadata *)meta;
+    if (md->type == FILE_SEG) {
+        hf3fs_utils->deregisterFileHandle(md->handle.fd);
+    } else if (md->type == DRAM_SEG) {
+        return NIXL_SUCCESS;
+    } else {
+        NIXL_LOG_AND_RETURN_IF_ERROR(NIXL_ERR_BACKEND, "Error - type not supported");
+    }
+    return NIXL_SUCCESS;
+}
+
+void nixlHf3fsEngine::cleanupIOList(nixlHf3fsBackendReqH *handle) const
+{
+    for (auto prev_io : handle->io_list) {
+        hf3fs_utils->destroyIOV(&prev_io->iov);
+        delete prev_io;
+    }
+
+    handle->io_list.clear();
+}
+
+void nixlHf3fsEngine::cleanupIOThread(nixlHf3fsBackendReqH *handle) const
+{
+    if (handle->io_status.thread != nullptr) {
+        handle->io_status.stop_thread = true;
+        handle->io_status.thread->join();
+
+        delete handle->io_status.thread;
+        handle->io_status.thread = nullptr;
+        handle->io_status.error_status = NIXL_SUCCESS;
+        handle->io_status.error_message = "";
+        handle->io_status.stop_thread = false;
+    }
+}
+
+nixl_status_t nixlHf3fsEngine::prepXfer (const nixl_xfer_op_t &operation,
+                                         const nixl_meta_dlist_t &local,
+                                         const nixl_meta_dlist_t &remote,
+                                         const std::string &remote_agent,
+                                         nixlBackendReqH* &handle,
+                                         const nixl_opt_b_args_t* opt_args) const
+{
+    nixlHf3fsBackendReqH *hf3fs_handle;
+    void                *addr = NULL;
+    size_t              size = 0;
+    size_t              offset = 0;
+    int                 buf_cnt  = local.descCount();
+    int                 file_cnt = remote.descCount();
+    nixl_status_t       nixl_err = NIXL_ERR_UNKNOWN;
+    const char          *nixl_mesg = nullptr;
+
+    // Determine which lists contain file/memory descriptors
+    const nixl_meta_dlist_t* file_list = nullptr;
+    const nixl_meta_dlist_t* mem_list = nullptr;
+    if (local.getType() == FILE_SEG) {
+        file_list = &local;
+        mem_list = &remote;
+    } else if (remote.getType() == FILE_SEG) {
+        file_list = &remote;
+        mem_list = &local;
+    } else {
+        NIXL_LOG_AND_RETURN_IF_ERROR(NIXL_ERR_INVALID_PARAM, "Error: No file descriptors");
+    }
+
+    if ((buf_cnt != file_cnt) ||
+        ((operation != NIXL_READ) && (operation != NIXL_WRITE)))  {
+        NIXL_LOG_AND_RETURN_IF_ERROR(NIXL_ERR_INVALID_PARAM,
+            "Error: Count mismatch or invalid operation selection");
+    }
+
+    hf3fs_handle = new nixlHf3fsBackendReqH();
+
+    bool is_read = (operation == NIXL_READ);
+
+    auto status = hf3fs_utils->createIOR(&hf3fs_handle->ior, file_cnt, is_read);
```

Comment:
As a perf improvement lets create IORs in a pool in future. This is creating IOR/treq.